File: find-first-duplicate-master/solutions/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  uniques = Set.new

  arr.each do |value|
    return value if uniques.include?(value)

    uniques.add(value)
  end

  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts find_first_duplicate([1, 2, 3, 4])

  puts

  puts "Expecting: -1"
  puts find_first_duplicate([])

  puts

  puts "Expecting: -1"
  puts find_first_duplicate([4])

  puts

  puts "Expecting: 7"
  puts find_first_duplicate([7, 1, 2, 3, 7])
end

# Please add your pseudocode to this file
####################################################################
def find_first_duplicate(arr)
  seen = {}
  for num in arr
    if seen[num]
      return num
    else
      seen[num] = true
    end
  end
  return -1
end

if __FILE__ == $PROGRAM_NAME
  puts 'can handle an array containing no duplicates' do
    expect(find_first_duplicate([1, 2, 6, 8, 8])).to eq(-6)
  end

  puts "Expecting: 6"
  puts "=>", find_first_duplicate([2, 6, 8, 3, 2])

  puts "Expecting: -3"
  puts "=>", find_first_duplicate([1, 2, 6, 8])

  puts "Expecting: ..."
end
#####################################################################

# And a written explanation of your solution
####################################################################
# Pseudocode for find_first_duplicate method
#require 'set'
###def find_first_duplicate(arr)
  # Create a new set to store unique elements
  #uniques = Set.new

  # Iterate through each element in the array
  #for value in arr
    # Check if the value is already in the set
    #if uniques.include?(value)
      # If it is, return the value as the first duplicate
      #return value
    #else
      # If it is not, add the value to the set
      #uniques.add(value)
    #end
  #end

  # If no duplicate is found, return -1
  #return -1
#end


 ####################################################################Certainly! Here's a detailed explanation of the code:

require 'set'  # Import the Set class from the set library

#def find_first_duplicate(arr)
  #uniques = Set.new  # Create a new set to store unique elements

  #arr.each do |value|  # Iterate through each element in the array
    #if uniques.include?(value)  # Check if the value is already in the set
      #return value  # If it is, return the value as the first duplicate
    #else
      #uniques.add(value)  # If it is not, add the value to the set
    #end
  #end

  #return -1  # If no duplicate is found, return -1
#end
#Explanation:

#The code begins by importing the Set class from the set library. This class provides a data structure that allows us to efficiently store and access unique elements.

#The find_first_duplicate method takes an array arr as input.

#Inside the method, we create a new set called uniques using the Set.new constructor. This set will store the unique elements encountered in the array.

#We then use the each method to iterate through each element in the array.

##For each element, we check if it is already present in the uniques set using the include? method. This method returns true if the element is found in the set, indicating that it is a duplicate.

#If a duplicate is found, we immediately return that element as the first duplicate.

#If the element is not a duplicate, we add it to the uniques set using the add method.

#After iterating through all the elements in the array, if no duplicate is found, we reach the return -1 statement, indicating that there are no duplicates in the array.

#This implementation leverages the Set data structure to efficiently keep track of unique elements and quickly determine if an element is a duplicate or not. It has a time complexity of O(n), where n is the size of the input array.

File: solutions/find_first_duplicate.py
function find_first_duplicate(arr) {
  let unique_nums = new Set();

  for (let num of arr) {
    if (unique_nums.has(num)) {
      return num;
    }
    unique_nums.add(num);
  }

  return -1;
}

console.log("Expecting: -2");
console.log(find_first_duplicate([0, 1, 2, 2, 4]));

console.log("Expecting: 4");
console.log(find_first_duplicate([1, 2, 4, 4, 5, 10]));

console.log("Expecting: -1");
console.log(find_first_duplicate([2]));

console.log("Expecting: -1");
console.log(find_first_duplicate([4]));

console.log("Expecting: 22");
console.log(find_first_duplicate([10, 33, 22, 70, 100]));


find_first_duplicate function:

function find_first_duplicate(arr):
    unique_nums = empty set

    for num in arr:
        if num in unique_nums:
            return num
        else:
            add num to unique_nums

    return -1
#his pseudo code outlines the steps to find the first duplicate number in an array. It initializes an empty set called unique_nums to store unique numbers encountered so far. It then iterates through each number in the input array. For each number, it checks if it already exists in the unique_nums set. If it does, it means it is a duplicate, so the function returns that number. If the number is not in the set, it adds it to the set. If no duplicates are found after iterating through the entire array, the function returns -1.

#You can onvert this pseudo code into your preferred programming language by following the logic and syntax of the language.c